src/ir/operation.h

```diff
@@ -0,0 +1,156 @@
+    const Operator* parent() const { return parent_; }
+    const NamedValueListMap results() const { return results_; }
+
+   private:
+    friend class OperatorBuilder;
+
+    Impl(const Operator* parent) : parent_(parent) {}
+
+    // The operator for which we are defining the implementation.
+    const Operator* parent_;
+    OperationList operations_;
+    // Maps the outputs of the operator to the corresponding value.
+    // Note that a result can have more than one value which is used
+    // to represent non-determinism.
+    NamedValueListMap results_;
+  };
+
+  const std::string& name() const { return name_; }
+  const DataDeclMap& inputs() const { return inputs_; }
+  const DataDeclMap& outputs() const { return outputs_; }
+  const Impl* impl() const { return impl_.get(); }
+
+ private:
+  friend class OperatorBuilder;
+
+  Operator(std::string name) : name_(std::move(name)), impl_(nullptr) {}
+
+  std::string name_;
+  DataDeclMap inputs_;
+  DataDeclMap outputs_;
+  // If the implementation is nullptr, then this is one of the basic operators
+  // like `+`, `-`, etc., that cannot be split up any further.
+  std::unique_ptr<Impl> impl_;
+};
+
+// An Operation represents a unit of execution.
+class Operation {
+ public:
+  Operation(const Operator* op, NamedValueMap inputs)
+      : op_(op), inputs_(std::move(inputs)) {
+    CHECK(op != nullptr);
+    CHECK(op->inputs().size() == inputs_.size());
+  }
+
+  const Operator& op() const { return *op_; }
+
+ private:
+  const Operator* op_;
+  NamedValueMap inputs_;
+};
+
+class OperatorBuilder {
+ public:
+  OperatorBuilder(std::string name)
+      : operator_(new Operator(std::move(name))) {}
+
+  OperatorBuilder& AddInput(std::string name,
+                            std::unique_ptr<types::Type> type) {
+    auto insertion_result =
+        operator_->inputs_.insert({std::move(name), std::move(type)});
+    CHECK(insertion_result.second);
+    return *this;
+  }
+
+  OperatorBuilder& AddOutput(std::string name,
+                             std::unique_ptr<types::Type> type) {
+    auto insertion_result =
+        operator_->outputs_.insert({std::move(name), std::move(type)});
+    CHECK(insertion_result.second);
+    return *this;
+  }
+
+  OperatorBuilder& AddOperation(const Operator* op, NamedValueMap inputs) {
+    GetImpl()->operations_.push_back(Operation(op, std::move(inputs)));
+    return *this;
+  }
+
+  OperatorBuilder& AddImplementation(
+      std::function<void(OperatorBuilder&, Operator*)> builder) {
+    builder(*this, operator_.get());
+    return *this;
+  }
+
+  OperatorBuilder& AddResult(absl::string_view name, Value output) {
+    CHECK(operator_->outputs_.find(name) != operator_->outputs_.end());
+    GetImpl()->results_[name].push_back(output);
+    return *this;
+  }
+
+  std::unique_ptr<Operator> build() { return std::move(operator_); }
+
+ private:
+  Operator::Impl* GetImpl() {
+    if (operator_->impl_ == nullptr) {
```

Comment:
This can be written as:

```C++
if (!operator_->impl_) {
```

Usually, I'm pretty creeped out by implicit conversions and avoid them, but I think it's better to rely upon the implicit conversion provided than to break encapsulation by assume that an absent managed pointer is represented with `nullptr` (although that is a correct assumption).